### ISSUE TITLE
DMA error after compiling on Pi 3

### ISSUE THREAD

**Issue by scramble45** [uunicorn/garage#2]
dmesg output while running test script.

````
[614980.948571] garage-door garage-door: error: failed to identify allocated channel
[614982.160449] bcm2835-dma 3f007000.dma: DMA transfer could not be terminated
[614982.160471] garage-door garage-door: error: failed to identify allocated channel
[614987.617608] bcm2835-dma 3f007000.dma: DMA transfer could not be terminated
[614987.617638] garage-door garage-door: error: failed to identify allocated channel
[615035.309546] bcm2835-dma 3f007000.dma: DMA transfer could not be terminated
[615035.309573] garage-door garage-door: error: failed to identify allocated channel
[615126.601112] bcm2835-dma 3f007000.dma: DMA transfer could not be terminated
[615126.601141] garage-door garage-door: error: failed to identify allocated channel
[615241.928753] bcm2835-dma 3f007000.dma: dma_pool_destroy 3f007000.dma, b6c1a000 busy
[615241.928795] Goodbye world.
[615255.076683] Allocated DMA channel 4
[615258.690896] bcm2835-dma 3f007000.dma: dma_pool_free 3f007000.dma, b6c1a000/f6c1a000 (bad dma)
[615258.691508] bcm2835-dma 3f007000.dma: DMA transfer could not be terminated
[615258.691886] garage-door garage-door: error: failed to identify allocated channel

````

**Comment by scramble45** [comment]
Recompiled tried again dmesg:
````
Detected hw channel 9.
[   63.842112] garage-door garage-door: error: failed to identify allocated channel
````

**Comment by uunicorn** [comment]
Basically, the idea behind the driver was this:
- Use a single DMA transaction for the whole door sequence: that should solve any timing issues caused by CPU load;
- Use PWM1 to generate the carrier (works in serializer mode, serializing 0x00000000 for "off" and 0xaaaaaaaa for the carrier, FIFO/DMA disabled);
- Use PWM2 to clock bits in the sequence (in M/N mode, FIFO/DMA enabled)
To achieve this, the driver must construct a long chain of DMA instructions which will interleave unconditional writes to PWM1 data register and writes to PWM2 FIFO register triggered by DREQ.

This has proven to work reliably on RPi2, but it turns out that bcm2835 dmaengine back-end lacks the implementation of interleaved xfers described above. At that point it was too late to give up, so using a lot of hackery I'm tring to figure out hardware channel number and manually program the DMA controller. First I'm allocating a DMA channel via public API, then I'm setting up a dummy xfer and trying to guess the allocated hardware channel by iterating over all DMA source address registers comparing them to my xfer source. This is extremely dodgy and assumes a lot about dmaengine backend. Apparently this hackery is broken on RPi 3 now and I wont bother trying to fix it.

TL;DR - if you want to make it right, you have two options: implement dmaengine_prep_interleaved_dma() for bcm2835 and remove all the hackery from this driver, or use a timer callback to clock the sequence bits. 

Personally I ended up running a trivial userspace program from my CGI script, which uses usleep() to clock the sequence bits instead of rebuilding the driver each the kernel gets upgraded. I'm using it every day for half a year now without any problem. Let me know if you need sources.

**Comment by scramble45** [comment]
Thanks for the detailed response. I will give this a shot on a Pi 2 and see if it works for me and the sources would be great. Thanks

**Comment by uunicorn** [comment]
Added pure userspace implementation.

**Comment by scramble45** [comment]
Yea that code is working well, seems to work on the pi3 from what I can tell with my scope. Thanks for publishing that.
